File: libs/counter/ios/CounterExample.swift
import SwiftUI
import WebKit

/**
 * Example SwiftUI view showing how to use the CounterBridge
 */
struct CounterExampleView: View {
    @StateObject private var viewModel = CounterViewModel()
    
    var body: some View {
        VStack(spacing: 0) {
            // Title
            Text("Counter Example")
                .font(.largeTitle)
                .padding()
            
            // Top part: WebView with blue border
            VStack {
                Text("WebView Counter")
                    .font(.headline)
                    .padding(.top)
                
                WebViewContainer(viewModel: viewModel)
                    .padding(10)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(
                RoundedRectangle(cornerRadius: 0)
                    .stroke(Color.blue, lineWidth: 4)
            )
            
            // Bottom part: Native component with red border
            VStack(spacing: 20) {
                Text("Native Counter")
                    .font(.headline)
                    .padding(.top)
                
                Text("Counter: \(viewModel.count)")
                    .font(.title)
                    .padding()
                
                HStack(spacing: 20) {
                    Button(action: {
                        viewModel.decrement()
                    }) {
                        Text("-")
                            .font(.title)
                            .frame(width: 50, height: 50)
                            .background(Color.red)
                            .foregroundColor(.white)
                            .cornerRadius(10)
                    }
                    
                    Button(action: {
                        viewModel.increment()
                    }) {
                        Text("+")
                            .font(.title)
                            .frame(width: 50, height: 50)
                            .background(Color.green)
                            .foregroundColor(.white)
                            .cornerRadius(10)
                    }
                }
                
                Button(action: {
                    viewModel.reset()
                }) {
                    Text("Reset")
                        .font(.headline)
                        .padding()
                        .background(Color.blue)
                        .foregroundColor(.white)
                        .cornerRadius(10)
                }
                .padding(.bottom)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(
                RoundedRectangle(cornerRadius: 0)
                    .stroke(Color.red, lineWidth: 4)
            )
        }
        .edgesIgnoringSafeArea(.bottom)
    }
}

/**
 * WebView container that integrates with CounterBridge
 */
struct WebViewContainer: UIViewRepresentable {
    @ObservedObject var viewModel: CounterViewModel
    
    func makeUIView(context: Context) -> WKWebView {
        let webView = WKWebView()
        webView.navigationDelegate = context.coordinator
        
        // Initialize the counter bridge
        viewModel.initializeBridge(webView: webView)
        
        // Load the web page that contains the counter
        if let url = URL(string: "http://localhost:3000") {
            webView.load(URLRequest(url: url))
        }
        
        return webView
    }
    
    func updateUIView(_ uiView: WKWebView, context: Context) {
        // No updates needed
    }
    
    func makeCoordinator() -> Coordinator {
        Coordinator(self)
    }
    
    class Coordinator: NSObject, WKNavigationDelegate {
        var parent: WebViewContainer
        
        init(_ parent: WebViewContainer) {
            self.parent = parent
        }
        
        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            // Initialize bridge after page load
            parent.viewModel.counterBridge?.initialize()
        }
    }
}

/**
 * ViewModel for the counter example
 */
class CounterViewModel: ObservableObject {
    @Published var count: Int = 0
    var counterBridge: CounterBridge?
    private var unsubscribe: (() -> Void)?
    
    func initializeBridge(webView: WKWebView) {
        // Create counter bridge
        counterBridge = CounterBridge(webView: webView)
        
        // Subscribe to counter changes
        unsubscribe = counterBridge?.subscribe { [weak self] newCount in
            DispatchQueue.main.async {
                self?.count = newCount
            }
        }
    }
    
    func increment() {
        counterBridge?.increment()
    }
    
    func decrement() {
        counterBridge?.decrement()
    }
    
    func reset() {
        counterBridge?.reset()
    }
    
    func setValue(_ value: Int) {
        counterBridge?.setValue(value: value)
    }
    
    deinit {
        // Clean up subscription when view model is deallocated
        unsubscribe?()
        counterBridge?.cleanup()
    }
}

/**
 * Example of how to integrate CounterBridge in an existing UIViewController
 */
class CounterExampleViewController: UIViewController, WKNavigationDelegate {
    private var webView: WKWebView!
    private var counterBridge: CounterBridge?
    private var counterLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Set up WebView
        webView = WKWebView(frame: CGRect(x: 0, y: 0, width: view.bounds.width, height: 300))
        webView.navigationDelegate = self
        view.addSubview(webView)
        
        // Set up counter label
        counterLabel = UILabel(frame: CGRect(x: 0, y: 320, width: view.bounds.width, height: 50))
        counterLabel.textAlignment = .center
        counterLabel.font = UIFont.systemFont(ofSize: 24)
        counterLabel.text = "Counter: 0"
        view.addSubview(counterLabel)
        
        // Load the web page
        if let url = URL(string: "http://localhost:3000") {
            webView.load(URLRequest(url: url))
        }
    }
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        // Initialize counter bridge
        counterBridge = CounterBridge(webView: webView)
        
        // Subscribe to counter changes
        counterBridge?.subscribe { [weak self] count in
            DispatchQueue.main.async {
                self?.counterLabel.text = "Counter: \(count)"
            }
        }
    }
    
    @objc func incrementCounter() {
        counterBridge?.increment()
    }
    
    @objc func decrementCounter() {
        counterBridge?.decrement()
    }
    
    @objc func resetCounter() {
        counterBridge?.reset()
    }
}

File: libs/counter/android/CounterExample.kt
package com.aha.counter.example

import android.os.Bundle
import android.view.View
import android.webkit.WebView
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.aha.counter.CounterBridge

/**
 * Example activity showing how to use the CounterBridge
 */
class CounterExampleActivity : AppCompatActivity() {
    
    private lateinit var webView: WebView
    private lateinit var counterBridge: CounterBridge
    private lateinit var counterValueText: TextView
    private lateinit var incrementButton: Button
    private lateinit var decrementButton: Button
    private lateinit var resetButton: Button
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        // Set up layout
        // Note: This is just an example, you would need to create the actual layout XML
        setContentView(R.layout.activity_counter_example)
        
        // Initialize WebView
        webView = findViewById(R.id.webview)
        webView.settings.javaScriptEnabled = true
        
        // Initialize counter bridge
        counterBridge = CounterBridge(webView)
        
        // Initialize UI elements
        counterValueText = findViewById(R.id.counter_value)
        incrementButton = findViewById(R.id.increment_button)
        decrementButton = findViewById(R.id.decrement_button)
        resetButton = findViewById(R.id.reset_button)
        
        // Set up button click listeners
        incrementButton.setOnClickListener {
            counterBridge.increment()
        }
        
        decrementButton.setOnClickListener {
            counterBridge.decrement()
        }
        
        resetButton.setOnClickListener {
            counterBridge.reset()
        }
        
        // Subscribe to counter changes
        counterBridge.subscribe { count ->
            // Update UI with new count
            runOnUiThread {
                counterValueText.text = count.toString()
            }
        }
        
        // Load the web page that contains the counter
        webView.loadUrl("http://localhost:3000")
        
        // Initialize the counter bridge after page load
        webView.webViewClient = object : android.webkit.WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                counterBridge.initialize()
            }
        }
    }
}

/**
 * Example of how to integrate CounterBridge in an existing activity
 */
fun integrateCounterBridgeExample(activity: AppCompatActivity, webView: WebView) {
    // Create counter bridge
    val counterBridge = CounterBridge(webView)
    
    // Initialize after page load
    webView.webViewClient = object : android.webkit.WebViewClient() {
        override fun onPageFinished(view: WebView?, url: String?) {
            super.onPageFinished(view, url)
            counterBridge.initialize()
        }
    }
    
    // Subscribe to counter changes
    val unsubscribe = counterBridge.subscribe { count ->
        // Do something with the count
        println("Counter value changed: $count")
    }
    
    // Example of incrementing the counter
    counterBridge.increment()
    
    // Example of setting a specific value
    counterBridge.setValue(10)
    
    // Example of unsubscribing when no longer needed
    // unsubscribe()
}